File: dameng/src/main/java/com/dbx/agent/dameng/DamengAgent.java
package com.dbx.agent.dameng;

import com.dbx.agent.BaseDatabaseAgent;
import com.dbx.agent.ColumnInfo;
import com.dbx.agent.ConnectParams;
import com.dbx.agent.DatabaseInfo;
import com.dbx.agent.ExecuteQueryOptions;
import com.dbx.agent.ForeignKeyInfo;
import com.dbx.agent.IndexInfo;
import com.dbx.agent.JdbcExecutor;
import com.dbx.agent.JdbcIdentifiers;
import com.dbx.agent.JsonRpcServer;
import com.dbx.agent.ObjectInfo;
import com.dbx.agent.ObjectSource;
import com.dbx.agent.QueryPageOptions;
import com.dbx.agent.QueryPageResult;
import com.dbx.agent.QueryResult;
import com.dbx.agent.TableInfo;
import com.dbx.agent.TriggerInfo;
import java.sql.Blob;
import java.sql.Clob;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLXML;
import java.sql.Types;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public final class DamengAgent extends BaseDatabaseAgent {
    private static final String AGENT_VERSION = "9999.06.04.1-fix-default";
    private static final Set<String> SYSTEM_USERS = Set.of(
        "SYS", "SYSAUDITOR", "SYSSSO", "CTISYS",
        "SYS_DBA", "_SYS_STATISTICS", "SYS_PHM"
    );

    private Connection connection;

    @Override
    public Connection getConnection() {
        return connection;
    }

    @Override
    public void connect(ConnectParams params) {
        uncheckedVoid(() -> {
            Class.forName("dm.jdbc.driver.DmDriver");
            connection = DriverManager.getConnection(buildUrl(params), params.getUsername(), params.getPassword());
        });
    }

    @Override
    public boolean testConnection(ConnectParams params) {
        return unchecked(() -> {
            Class.forName("dm.jdbc.driver.DmDriver");
            try (Connection conn = DriverManager.getConnection(buildUrl(params), params.getUsername(), params.getPassword())) {
                return conn.isValid(5);
            }
        });
    }

    @Override
    public List<DatabaseInfo> listDatabases() {
        return unchecked(() -> {
            List<DatabaseInfo> result = new ArrayList<>();
            String placeholders = String.join(",", SYSTEM_USERS.stream().map(user -> "?").toList());
            String sql = "SELECT USERNAME FROM ALL_USERS WHERE USERNAME NOT IN (" + placeholders + ") ORDER BY USERNAME";
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                int index = 1;
                for (String user : SYSTEM_USERS) {
                    stmt.setString(index++, user);
                }
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new DatabaseInfo(rs.getString(1)));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<String> listSchemas() {
        return unchecked(() -> {
            List<String> result = new ArrayList<>();
            String sql = "SELECT DISTINCT OWNER FROM ALL_OBJECTS ORDER BY OWNER";
            try (java.sql.Statement stmt = requireConnected().createStatement();
                 ResultSet rs = stmt.executeQuery(sql)) {
                while (rs.next()) {
                    result.add(rs.getString(1));
                }
            }
            return result;
        });
    }

    @Override
    public List<TableInfo> listTables(String schema) {
        return unchecked(() -> {
            List<TableInfo> result = new ArrayList<>();
            String sql = """
                SELECT TABLE_NAME, 'TABLE' AS TABLE_TYPE, COMMENTS
                FROM ALL_TAB_COMMENTS
                WHERE OWNER = ? AND TABLE_TYPE = 'TABLE'
                UNION ALL
                SELECT TABLE_NAME, 'VIEW' AS TABLE_TYPE, COMMENTS
                FROM ALL_TAB_COMMENTS
                WHERE OWNER = ? AND TABLE_TYPE = 'VIEW'
                ORDER BY 1
                """.stripIndent().trim();
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, schema);
                stmt.setString(2, schema);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new TableInfo(rs.getString(1), rs.getString(2), rs.getString(3)));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<ObjectInfo> listObjects(String schema) {
        return unchecked(() -> {
            List<ObjectInfo> result = new ArrayList<>();
            for (TableInfo table : listTables(schema)) {
                result.add(new ObjectInfo(table.getName(), table.getTable_type(), schema, table.getComment()));
            }
            String sql = """
                SELECT OBJECT_NAME, OBJECT_TYPE FROM ALL_OBJECTS
                WHERE OWNER = ? AND OBJECT_TYPE IN ('PROCEDURE', 'FUNCTION')
                ORDER BY CASE OBJECT_TYPE WHEN 'PROCEDURE' THEN 0 ELSE 1 END, OBJECT_NAME
                """.stripIndent().trim();
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, schema);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new ObjectInfo(rs.getString(1), rs.getString(2), schema, null));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public ObjectSource getObjectSource(String schema, String name, String objectType) {
        return unchecked(() -> {
            String dbmsType = switch (objectType.toUpperCase(Locale.ROOT)) {
                case "VIEW" -> "VIEW";
                case "PROCEDURE" -> "PROCEDURE";
                case "FUNCTION" -> "FUNCTION";
                default -> throw new IllegalArgumentException("Unsupported object type: " + objectType);
            };
            String source;
            String sql = "SELECT DBMS_METADATA.GET_DDL(?, ?, ?) FROM DUAL";
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, dbmsType);
                stmt.setString(2, name);
                stmt.setString(3, schema);
                try (ResultSet rs = stmt.executeQuery()) {
                    source = rs.next() ? coalesce(rs.getString(1)) : "";
                }
            }
            return new ObjectSource(name, objectType, schema, source);
        });
    }

    @Override
    public String getTableDdl(String schema, String table) {
        return unchecked(() -> {
            String sql = "SELECT DBMS_METADATA.GET_DDL(?, ?, ?) FROM DUAL";
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, "TABLE");
                stmt.setString(2, table);
                stmt.setString(3, schema);
                try (ResultSet rs = stmt.executeQuery()) {
                    if (rs.next()) {
                        return coalesce(rs.getString(1));
                    }
                }
            }
            throw new IllegalArgumentException("Table not found: " + schema + "." + table);
        });
    }

    @Override
    public List<ColumnInfo> getColumns(String schema, String table) {
        return unchecked(() -> {
            Set<String> pkColumns = new java.util.HashSet<>();
            String pkSql = """
                SELECT cols.COLUMN_NAME FROM ALL_CONS_COLUMNS cols
                JOIN ALL_CONSTRAINTS cons ON cols.CONSTRAINT_NAME = cons.CONSTRAINT_NAME AND cols.OWNER = cons.OWNER
                WHERE cons.CONSTRAINT_TYPE = 'P' AND cons.OWNER = ? AND cons.TABLE_NAME = ?
                """.stripIndent().trim();
            try (PreparedStatement stmt = requireConnected().prepareStatement(pkSql)) {
                stmt.setString(1, schema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        pkColumns.add(rs.getString(1));
                    }
                }
            }

            List<ColumnInfo> result = new ArrayList<>();
            // DATA_DEFAULT is a LONG column — it must be selected first and read first
            // in JDBC, otherwise the data is truncated.
            String colSql = """
                SELECT c.DATA_DEFAULT,
                    c.COLUMN_NAME,
                    c.DATA_TYPE,
                    c.NULLABLE,
                    c.DATA_PRECISION,
                    c.DATA_SCALE,
                    c.DATA_LENGTH,
                    c.CHAR_LENGTH,
                    cc.COMMENTS
                FROM ALL_TAB_COLUMNS c
                LEFT JOIN ALL_COL_COMMENTS cc
                    ON cc.OWNER = c.OWNER
                    AND cc.TABLE_NAME = c.TABLE_NAME
                    AND cc.COLUMN_NAME = c.COLUMN_NAME
                WHERE c.OWNER = ? AND c.TABLE_NAME = ?
                ORDER BY c.COLUMN_ID
                """.stripIndent().trim();
            try (PreparedStatement stmt = requireConnected().prepareStatement(colSql)) {
                stmt.setString(1, schema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        // DATA_DEFAULT is a LONG — must be read first, before all other columns.
                        String dataDefault = readLongColumn(rs, "DATA_DEFAULT");
                        String name = rs.getString("COLUMN_NAME");
                        String baseType = rs.getString("DATA_TYPE");
                        Integer numPrec = intObject(rs, "DATA_PRECISION");
                        Integer numScale = intObject(rs, "DATA_SCALE");
                        Integer dataLen = intObject(rs, "DATA_LENGTH");
                        Integer charLen = intObject(rs, "CHAR_LENGTH");
                        String dataType = formatDataType(baseType, numPrec, numScale, dataLen, charLen);

                        result.add(new ColumnInfo(
                            name,
                            dataType,
                            "Y".equals(rs.getString("NULLABLE")),
                            dataDefault,
                            pkColumns.contains(name),
                            null,
                            rs.getString("COMMENTS"),
                            numPrec,
                            numScale,
                            charLen
                        ));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<IndexInfo> listIndexes(String schema, String table) {
        return unchecked(() -> {
            List<IndexInfo> result = new ArrayList<>();
            String sql = """
                SELECT i.INDEX_NAME,
                    LISTAGG(ic.COLUMN_NAME, ',') WITHIN GROUP (ORDER BY ic.COLUMN_POSITION) AS COLUMNS,
                    i.UNIQUENESS,
                    CASE WHEN c.CONSTRAINT_TYPE = 'P' THEN 1 ELSE 0 END AS IS_PK,
                    i.INDEX_TYPE
                FROM ALL_INDEXES i
                JOIN ALL_IND_COLUMNS ic ON i.INDEX_NAME = ic.INDEX_NAME AND i.OWNER = ic.INDEX_OWNER AND i.TABLE_OWNER = ic.TABLE_OWNER
                LEFT JOIN ALL_CONSTRAINTS c ON i.INDEX_NAME = c.INDEX_NAME AND i.TABLE_OWNER = c.OWNER
                    AND c.CONSTRAINT_TYPE = 'P'
                WHERE i.TABLE_OWNER = ? AND i.TABLE_NAME = ?
                GROUP BY i.INDEX_NAME, i.UNIQUENESS, c.CONSTRAINT_TYPE, i.INDEX_TYPE
                ORDER BY i.INDEX_NAME
                """.stripIndent().trim();
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, schema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new IndexInfo(
                            rs.getString(1),
                            splitNonEmpty(coalesce(rs.getString(2)), ","),
                            "UNIQUE".equals(rs.getString(3)),
                            "1".equals(rs.getString(4)),
                            null,
                            rs.getString(5),
                            null,
                            null
                        ));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<ForeignKeyInfo> listForeignKeys(String schema, String table) {
        return unchecked(() -> {
            List<ForeignKeyInfo> result = new ArrayList<>();
            String sql = """
                SELECT c.CONSTRAINT_NAME, cc.COLUMN_NAME, rc.TABLE_NAME, rcc.COLUMN_NAME
                FROM ALL_CONSTRAINTS c
                JOIN ALL_CONS_COLUMNS cc ON c.CONSTRAINT_NAME = cc.CONSTRAINT_NAME AND c.OWNER = cc.OWNER
                JOIN ALL_CONSTRAINTS rc ON c.R_CONSTRAINT_NAME = rc.CONSTRAINT_NAME AND c.R_OWNER = rc.OWNER
                JOIN ALL_CONS_COLUMNS rcc ON rc.CONSTRAINT_NAME = rcc.CONSTRAINT_NAME AND rc.OWNER = rcc.OWNER
                WHERE c.CONSTRAINT_TYPE = 'R' AND c.OWNER = ? AND c.TABLE_NAME = ?
                ORDER BY c.CONSTRAINT_NAME
                """.stripIndent().trim();
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, schema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new ForeignKeyInfo(
                            rs.getString(1),
                            rs.getString(2),
                            rs.getString(3),
                            rs.getString(4)
                        ));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<TriggerInfo> listTriggers(String schema, String table) {
        return unchecked(() -> {
            List<TriggerInfo> result = new ArrayList<>();
            String sql = """
                SELECT TRIGGER_NAME, TRIGGERING_EVENT, '' AS TRIGGER_TYPE
                FROM ALL_TRIGGERS
                WHERE OWNER = ? AND TABLE_NAME = ?
                ORDER BY TRIGGER_NAME
                """.stripIndent().trim();
            try (PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, schema);
                stmt.setString(2, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new TriggerInfo(rs.getString(1), rs.getString(2), rs.getString(3)));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public QueryResult executeQuery(String sql, String schema, ExecuteQueryOptions options) {
        return JdbcExecutor.INSTANCE.execute(
            requireConnected(),
            sql,
            schema,
            this::setSchemaSQL,
            options.getMaxRows(),
            options.getFetchSize(),
            this::stringResultValue
        );
    }

    @Override
    public QueryPageResult executeQueryPage(String sql, String schema, QueryPageOptions options) {
        return JdbcExecutor.INSTANCE.executePage(
            requireConnected(),
            sql,
            schema,
            this::setSchemaSQL,
            options,
            this::stringResultValue
        );
    }

    @Override
    public String setSchemaSQL(String schema) {
        return "SET SCHEMA " + JdbcIdentifiers.INSTANCE.doubleQuote(schema);
    }

    @Override
    public void disconnect() {
        uncheckedVoid(() -> {
            if (connection != null) {
                connection.close();
            }
            connection = null;
        });
    }

    private Object stringResultValue(ResultSet rs, int index, int sqlType) {
        return unchecked(() -> {
            Object value = switch (sqlType) {
                case Types.BIGINT -> rs.getLong(index);
                case Types.INTEGER, Types.SMALLINT, Types.TINYINT -> rs.getInt(index);
                case Types.FLOAT, Types.REAL -> rs.getFloat(index);
                case Types.DOUBLE -> rs.getDouble(index);
                case Types.DECIMAL, Types.NUMERIC -> rs.getBigDecimal(index);
                case Types.BOOLEAN, Types.BIT -> rs.getBoolean(index);
                case Types.CHAR, Types.VARCHAR, Types.LONGVARCHAR,
                    Types.NCHAR, Types.NVARCHAR, Types.LONGNVARCHAR,
                    Types.CLOB, Types.NCLOB -> rs.getString(index);
                case Types.BINARY, Types.VARBINARY, Types.LONGVARBINARY,
                    Types.BLOB -> bytesToHex(rs.getBytes(index));
                case Types.SQLXML -> sqlXmlToString(rs.getSQLXML(index));
                default -> normalizeResultValue(rs.getObject(index));
            };
            return rs.wasNull() ? null : value;
        });
    }

    private static Object normalizeResultValue(Object value) {
        return switch (value) {
            case null -> null;
            case Clob clob -> unchecked(() -> clob.getSubString(1, Math.toIntExact(clob.length())));
            case Blob blob -> unchecked(() -> bytesToHex(blob.getBytes(1, Math.toIntExact(blob.length()))));
            case SQLXML sqlxml -> unchecked(sqlxml::getString);
            case byte[] bytes -> bytesToHex(bytes);
            default -> value instanceof Number || value instanceof Boolean ? value : value.toString();
        };
    }

    private static String sqlXmlToString(SQLXML value) {
        return value == null ? null : unchecked(value::getString);
    }

    private static String bytesToHex(byte[] bytes) {
        if (bytes == null) {
            return null;
        }
        StringBuilder result = new StringBuilder(bytes.length * 2 + 2);
        result.append("0x");
        for (byte b : bytes) {
            result.append(Character.forDigit((b >> 4) & 0xF, 16));
            result.append(Character.forDigit(b & 0xF, 16));
        }
        return result.toString();
    }

    private static String buildUrl(ConnectParams params) {
        String database = params.getDatabase() == null ? "" : params.getDatabase().trim();
        String suffix = database.isEmpty() ? "" : "/" + database;
        return "jdbc:dm://" + params.getHost() + ":" + params.getPort() + suffix;
    }

    private static String formatDataType(String base, Integer numPrec, Integer numScale, Integer dataLen, Integer charLen) {
        return switch (base.toUpperCase(Locale.ROOT)) {
            case "VARCHAR2", "NVARCHAR2", "VARCHAR", "CHAR", "NCHAR" -> {
                Integer length = charLen != null ? charLen : dataLen;
                yield length != null ? base + "(" + length + ")" : base;
            }
            case "NUMBER", "NUMERIC", "DECIMAL" -> {
                if (numPrec != null && numScale != null && numScale > 0) {
                    yield base + "(" + numPrec + "," + numScale + ")";
                }
                yield numPrec != null && numPrec > 0 ? base + "(" + numPrec + ")" : base;
            }
            case "RAW" -> dataLen != null ? "RAW(" + dataLen + ")" : "RAW";
            default -> base;
        };
    }

    private static Integer intObject(ResultSet rs, String column) throws Exception {
        Object value = rs.getObject(column);
        return value == null ? null : ((Number) value).intValue();
    }

    // DATA_DEFAULT is stored as a LONG column in Oracle/Dameng. JDBC requires LONG
    // columns to be read before other columns. We also try getCharacterStream as a
    // fallback because some drivers don't support getString on LONG columns.
    private static String readLongColumn(ResultSet rs, String column) throws Exception {
        try {
            String value = rs.getString(column);
            if (value != null) {
                return value;
            }
        } catch (Exception ignored) {
        }
        try (java.io.Reader reader = rs.getCharacterStream(column)) {
            if (reader == null) {
                return null;
            }
            StringBuilder sb = new StringBuilder();
            char[] buf = new char[4096];
            int n;
            while ((n = reader.read(buf)) != -1) {
                sb.append(buf, 0, n);
            }
            return sb.toString();
        }
    }

    private static List<String> splitNonEmpty(String value, String delimiter) {
        List<String> result = new ArrayList<>();
        Arrays.stream(value.split(delimiter))
            .filter(part -> !part.isEmpty())
            .forEach(result::add);
        return result;
    }

    private static String coalesce(String value) {
        return value == null ? "" : value;
    }

    public static void main(String[] args) {
        new JsonRpcServer(new DamengAgent()).run();
    }
}

File: dameng/src/main/kotlin/com/dbx/agent/dameng/DamengAgent.kt
package com.dbx.agent.dameng

import com.dbx.agent.*
import java.sql.Connection
import java.sql.DriverManager
import java.sql.ResultSet
import java.sql.Types

class DamengAgent : DatabaseAgent {

    private var connection: Connection? = null

    override fun getConnection(): Connection? = connection
    companion object {
        private const val MAX_ROWS = 10000

        private val SYSTEM_USERS = setOf(
            "SYS", "SYSAUDITOR", "SYSSSO", "CTISYS",
            "SYS_DBA", "_SYS_STATISTICS", "SYS_PHM"
        )

        private val QUERY_PREFIXES = listOf("SELECT", "WITH", "SHOW", "DESCRIBE", "EXPLAIN")
    }

    override fun connect(params: ConnectParams) {
        Class.forName("dm.jdbc.driver.DmDriver")
        val url = "jdbc:dm://${params.host}:${params.port}/${params.database}"
        connection = DriverManager.getConnection(url, params.username, params.password)
    }

    override fun testConnection(params: ConnectParams): Boolean {
        Class.forName("dm.jdbc.driver.DmDriver")
        val url = "jdbc:dm://${params.host}:${params.port}/${params.database}"
        DriverManager.getConnection(url, params.username, params.password).use { conn ->
            return conn.isValid(5)
        }
    }

    override fun listDatabases(): List<DatabaseInfo> {
        val conn = requireConnection()
        val placeholders = SYSTEM_USERS.joinToString(",") { "?" }
        val sql = "SELECT USERNAME FROM ALL_USERS WHERE USERNAME NOT IN ($placeholders) ORDER BY USERNAME"
        conn.prepareStatement(sql).use { stmt ->
            SYSTEM_USERS.forEachIndexed { index, name ->
                stmt.setString(index + 1, name)
            }
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(DatabaseInfo(rs.getString(1)))
                    }
                }
            }
        }
    }

    override fun listSchemas(): List<String> {
        return listDatabases().map { it.name }
    }

    override fun listTables(schema: String): List<TableInfo> {
        val conn = requireConnection()
        val sql = """
            SELECT TABLE_NAME, 'TABLE' AS TABLE_TYPE FROM ALL_TABLES WHERE OWNER = ?
            UNION ALL
            SELECT VIEW_NAME, 'VIEW' FROM ALL_VIEWS WHERE OWNER = ?
            ORDER BY 1
        """.trimIndent()
        conn.prepareStatement(sql).use { stmt ->
            stmt.setString(1, schema)
            stmt.setString(2, schema)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(TableInfo(
                            name = rs.getString(1),
                            table_type = rs.getString(2)
                        ))
                    }
                }
            }
        }
    }

    override fun getColumns(schema: String, table: String): List<ColumnInfo> {
        val conn = requireConnection()

        // Fetch primary key columns
        val pkColumns = mutableSetOf<String>()
        val pkSql = """
            SELECT cols.COLUMN_NAME FROM ALL_CONS_COLUMNS cols
            JOIN ALL_CONSTRAINTS cons ON cols.CONSTRAINT_NAME = cons.CONSTRAINT_NAME AND cols.OWNER = cons.OWNER
            WHERE cons.CONSTRAINT_TYPE = 'P' AND cons.OWNER = ? AND cons.TABLE_NAME = ?
        """.trimIndent()
        conn.prepareStatement(pkSql).use { stmt ->
            stmt.setString(1, schema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                while (rs.next()) {
                    pkColumns.add(rs.getString(1))
                }
            }
        }

        // Fetch column metadata
        val colSql = """
            SELECT COLUMN_NAME, DATA_TYPE, NULLABLE, DATA_PRECISION, DATA_SCALE, DATA_LENGTH, CHAR_LENGTH
            FROM ALL_TAB_COLUMNS
            WHERE OWNER = ? AND TABLE_NAME = ?
            ORDER BY COLUMN_ID
        """.trimIndent()
        conn.prepareStatement(colSql).use { stmt ->
            stmt.setString(1, schema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        val name = rs.getString("COLUMN_NAME")
                        val baseType = rs.getString("DATA_TYPE")
                        val numPrec = rs.getObject("DATA_PRECISION")?.let { (it as Number).toInt() }
                        val numScale = rs.getObject("DATA_SCALE")?.let { (it as Number).toInt() }
                        val dataLen = rs.getObject("DATA_LENGTH")?.let { (it as Number).toInt() }
                        val charLen = rs.getObject("CHAR_LENGTH")?.let { (it as Number).toInt() }
                        val dataType = formatDataType(baseType, numPrec, numScale, dataLen, charLen)

                        add(ColumnInfo(
                            name = name,
                            data_type = dataType,
                            is_nullable = rs.getString("NULLABLE") == "Y",
                            column_default = null,
                            is_primary_key = pkColumns.contains(name),
                            extra = null,
                            comment = null,
                            numeric_precision = numPrec,
                            numeric_scale = numScale,
                            character_maximum_length = charLen
                        ))
                    }
                }
            }
        }
    }

    override fun listIndexes(schema: String, table: String): List<IndexInfo> {
        val conn = requireConnection()
        val sql = """
            SELECT i.INDEX_NAME,
                LISTAGG(ic.COLUMN_NAME, ',') WITHIN GROUP (ORDER BY ic.COLUMN_POSITION) AS COLUMNS,
                i.UNIQUENESS,
                CASE WHEN c.CONSTRAINT_TYPE = 'P' THEN 1 ELSE 0 END AS IS_PK,
                i.INDEX_TYPE
            FROM ALL_INDEXES i
            JOIN ALL_IND_COLUMNS ic ON i.INDEX_NAME = ic.INDEX_NAME AND i.OWNER = ic.INDEX_OWNER AND i.TABLE_OWNER = ic.TABLE_OWNER
            LEFT JOIN ALL_CONSTRAINTS c ON i.INDEX_NAME = c.INDEX_NAME AND i.TABLE_OWNER = c.OWNER
                AND c.CONSTRAINT_TYPE = 'P'
            WHERE i.TABLE_OWNER = ? AND i.TABLE_NAME = ?
            GROUP BY i.INDEX_NAME, i.UNIQUENESS, c.CONSTRAINT_TYPE, i.INDEX_TYPE
            ORDER BY i.INDEX_NAME
        """.trimIndent()
        conn.prepareStatement(sql).use { stmt ->
            stmt.setString(1, schema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        val colsStr = rs.getString(2) ?: ""
                        add(IndexInfo(
                            name = rs.getString(1),
                            columns = colsStr.split(",").filter { it.isNotEmpty() },
                            is_unique = rs.getString(3) == "UNIQUE",
                            is_primary = rs.getString(4) == "1",
                            filter = null,
                            index_type = rs.getString(5)
                        ))
                    }
                }
            }
        }
    }

    override fun listForeignKeys(schema: String, table: String): List<ForeignKeyInfo> {
        val conn = requireConnection()
        val sql = """
            SELECT c.CONSTRAINT_NAME, cc.COLUMN_NAME, rc.TABLE_NAME, rcc.COLUMN_NAME
            FROM ALL_CONSTRAINTS c
            JOIN ALL_CONS_COLUMNS cc ON c.CONSTRAINT_NAME = cc.CONSTRAINT_NAME AND c.OWNER = cc.OWNER
            JOIN ALL_CONSTRAINTS rc ON c.R_CONSTRAINT_NAME = rc.CONSTRAINT_NAME AND c.R_OWNER = rc.OWNER
            JOIN ALL_CONS_COLUMNS rcc ON rc.CONSTRAINT_NAME = rcc.CONSTRAINT_NAME AND rc.OWNER = rcc.OWNER
            WHERE c.CONSTRAINT_TYPE = 'R' AND c.OWNER = ? AND c.TABLE_NAME = ?
            ORDER BY c.CONSTRAINT_NAME
        """.trimIndent()
        conn.prepareStatement(sql).use { stmt ->
            stmt.setString(1, schema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(ForeignKeyInfo(
                            name = rs.getString(1),
                            column = rs.getString(2),
                            ref_table = rs.getString(3),
                            ref_column = rs.getString(4)
                        ))
                    }
                }
            }
        }
    }

    override fun listTriggers(schema: String, table: String): List<TriggerInfo> {
        val conn = requireConnection()
        val sql = """
            SELECT TRIGGER_NAME, TRIGGERING_EVENT, TRIGGER_TYPE
            FROM ALL_TRIGGERS
            WHERE OWNER = ? AND TABLE_NAME = ?
            ORDER BY TRIGGER_NAME
        """.trimIndent()
        conn.prepareStatement(sql).use { stmt ->
            stmt.setString(1, schema)
            stmt.setString(2, table)
            stmt.executeQuery().use { rs ->
                return buildList {
                    while (rs.next()) {
                        add(TriggerInfo(
                            name = rs.getString(1),
                            event = rs.getString(2),
                            timing = rs.getString(3)
                        ))
                    }
                }
            }
        }
    }

    override fun executeQuery(sql: String, schema: String?): QueryResult {
        val conn = requireConnection()
        val trimmedSql = sql.trim().trimEnd(';')
        val upperSql = trimmedSql.uppercase().trimStart()

        // Translate transaction control to JDBC calls
        if (upperSql == "BEGIN" || upperSql == "BEGIN TRANSACTION") {
            val start = System.currentTimeMillis()
            conn.autoCommit = false
            return QueryResult(emptyList(), emptyList(), 0, System.currentTimeMillis() - start)
        }
        if (upperSql == "COMMIT") {
            val start = System.currentTimeMillis()
            conn.commit()
            conn.autoCommit = true
            return QueryResult(emptyList(), emptyList(), 0, System.currentTimeMillis() - start)
        }
        if (upperSql == "ROLLBACK") {
            val start = System.currentTimeMillis()
            conn.rollback()
            conn.autoCommit = true
            return QueryResult(emptyList(), emptyList(), 0, System.currentTimeMillis() - start)
        }

        // Set schema if provided
        if (!schema.isNullOrBlank()) {
            conn.createStatement().use { stmt ->
                stmt.execute("SET SCHEMA \"$schema\"")
            }
        }

        val startTime = System.currentTimeMillis()
        val isQuery = QUERY_PREFIXES.any { upperSql.startsWith(it) }

        if (isQuery) {
            conn.createStatement().use { stmt ->
                stmt.executeQuery(trimmedSql).use { rs ->
                    val meta = rs.metaData
                    val colCount = meta.columnCount
                    val columns = (1..colCount).map { meta.getColumnLabel(it) }
                    val rows = mutableListOf<List<Any?>>()
                    while (rs.next() && rows.size < MAX_ROWS) {
                        val row = (1..colCount).map { i ->
                            val value = rs.getObject(i)
                            if (rs.wasNull()) null else value?.toString()
                        }
                        rows.add(row)
                    }
                    val elapsed = System.currentTimeMillis() - startTime
                    return QueryResult(
                        columns = columns,
                        rows = rows,
                        affected_rows = 0,
                        execution_time_ms = elapsed,
                        truncated = rows.size >= MAX_ROWS
                    )
                }
            }
        } else {
            conn.createStatement().use { stmt ->
                val affected = stmt.executeUpdate(trimmedSql)
                val elapsed = System.currentTimeMillis() - startTime
                return QueryResult(
                    columns = emptyList(),
                    rows = emptyList(),
                    affected_rows = affected.toLong(),
                    execution_time_ms = elapsed,
                    truncated = false
                )
            }
        }
    }

    override fun disconnect() {
        connection?.close()
        connection = null
    }

    private fun requireConnection(): Connection {
        return connection ?: throw IllegalStateException("Not connected")
    }

    private fun formatDataType(
        base: String,
        numPrec: Int?,
        numScale: Int?,
        dataLen: Int?,
        charLen: Int?
    ): String {
        return when (base.uppercase()) {
            "VARCHAR2", "NVARCHAR2", "VARCHAR", "CHAR", "NCHAR" -> {
                val len = charLen ?: dataLen
                if (len != null) "$base($len)" else base
            }
            "NUMBER", "NUMERIC", "DECIMAL" -> {
                when {
                    numPrec != null && numScale != null && numScale > 0 -> "$base($numPrec,$numScale)"
                    numPrec != null && numPrec > 0 -> "$base($numPrec)"
                    else -> base
                }
            }
            "RAW" -> {
                if (dataLen != null) "RAW($dataLen)" else "RAW"
            }
            else -> base
        }
    }
}

fun main() {
    JsonRpcServer(DamengAgent()).run()
}
